torch/_inductor/codegen/wrapper.py

```diff
@@ -2036,6 +2036,77 @@ def codegen_while_loop(self, while_loop):
         )
         self.writeline(ExitSubgraphLine(self))
 
+    def codegen_sequential_scan(self, sequential_scan):
+        from torch._higher_order_ops.scan import _extract_carry_and_out
+
+        name = sequential_scan.get_name()
+        outer_init = [buf.codegen_reference() for buf in sequential_scan.init]
+        outer_xs = [buf.codegen_reference() for buf in sequential_scan.xs]
+        outer_additional_inputs = [
+            buf.codegen_reference() for buf in sequential_scan.additional_inputs
+        ]
+        iter_idx_expr_str = str(sequential_scan.iter_idx_expr)
+        dim = sequential_scan.dim
+
+        self.writeline(f"{name} = [None] * {len(sequential_scan.outputs)}")
+        outer_outputs = [f"{name}[{i}]" for i in range(len(sequential_scan.outputs))]
+        outer_carry_out, outer_ys_out = _extract_carry_and_out(
+            outer_outputs, len(outer_init)
+        )
+        # Initialize the carry output to be a clone of init.
+        outer_carry_outs = []
+        inner_carry_outs = []
+        self.writeline(f"inner_carry_out = [None for i in range({len(outer_init)})]")
+        for i, init in enumerate(
+            _extract_carry_and_out(sequential_scan.outputs, len(outer_init))[0]
+        ):
+            self.writeline(f"inner_carry_out[{i}] = {outer_init[i]}")
+            inner_carry_outs.append(f"inner_carry_out[{i}]")
+
+        # Initialize the ys_output to an empty buffer.
+        outer_ys_outs = []
+        self.writeline(f"outer_ys_out = [None for i in range({len(outer_ys_out)})]")
+        for i, ys_out in enumerate(
+            _extract_carry_and_out(sequential_scan.outputs, len(outer_init))[1]
+        ):
+            self.writeline(
+                self.make_allocation(
+                    f"outer_ys_out[{i}]",
+                    ys_out.get_device(),
+                    ys_out.get_dtype(),
+                    ys_out.get_size(),
+                    ys_out.get_stride(),
+                )
+            )
+            self.writeline(f"{name}[{i+len(outer_carry_out)}] = outer_ys_out[{i}]")
+            outer_ys_outs.append(f"outer_ys_out[{i}]")
+
+        self.writeline(f"scan_length = {outer_xs[0]}.shape[{dim}]")
+        self.writeline(f"dim = {sequential_scan.dim}")
+        self.writeline(f"reverse = {sequential_scan.reverse}")
```

Comment:
I think `reverse` should just be handled at lowering time. Is there any reason for the reverse to persist through the lowering phase into the codegen?

Reply:
Right now all the strides and shapes are represented as a symbol (e.g. in the example s3) and we initialize the symbol based on the value of reverse`s3 = i if not reverse else scan_length - i - 1` If we're to handle at lowering time, we don't need to do the assignment for s3, but every tensor would take a simbolic shape like `scan_length - s3 -1`. Like in the example, the kernel call looks like:
```
s3 = idx
...
triton_poi_fused_clone_sin_3.run(... s2, scan_length - s3 -1, s2, grid=grid(s2), stream=stream0)
```
This causes repeated computation of the expression scan_length - s3 -1. By also passing in reverse call, we can initialize the s3 in one place, which saves cpu opertions and makes the output code easier to read.